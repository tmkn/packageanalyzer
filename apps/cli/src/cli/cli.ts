import { Cli, Builtins } from "clipanion";

import { AnalyzeCommand } from "./analyzeCommand.js";
import { UpdateInfoCommand } from "./updateInfoCommand.js";
import { DownloadCommand } from "./downloadCommand.js";
import { LoopsCommand } from "./loopsCommand.js";
import { TreeCommand } from "./treeCommand.js";
import { LicenseCheckCommand } from "./licenseCommand.js";
import { ReportCommand } from "./reportCommand.js";
import { DependencyDumperCommand } from "./dependencyDumpCommand.js";
import { TestCommand } from "./testCommand.js";
import { DiffCommand } from "./diffCommand.js";
import { LintCommand } from "./lintCommand.js";
import { getVersion } from "../../../../packages/node/src/common.js";

export const cli = new Cli({
    binaryLabel: `packageanalyzer`,
    binaryName: `pkga`,
    binaryVersion: getVersion()
});

//standard commands
cli.register(ReportCommand);
cli.register(AnalyzeCommand);
cli.register(UpdateInfoCommand);
cli.register(DownloadCommand);
cli.register(LoopsCommand);
cli.register(TreeCommand);
cli.register(LicenseCheckCommand);
cli.register(DiffCommand);
cli.register(LintCommand);

//development niche commands
cli.register(DependencyDumperCommand);

cli.register(TestCommand);

//built in commands
cli.register(Builtins.HelpCommand);
cli.register(Builtins.VersionCommand);
