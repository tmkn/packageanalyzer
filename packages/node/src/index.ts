export { Visitor, getPackageVersionfromString } from "../../shared/src/visitors/visitor.js";
export type { PackageVersion } from "../../shared/src/visitors/visitor.js";
export { Package } from "../../shared/src/package/package.js";
export { npmOnline, OnlinePackageProvider } from "./providers/online.js";
export type { IPackageJsonProvider } from "../../shared/src/providers/provider.js";
export { OraLogger } from "./loggers/OraLogger.js";
export type { ILogger } from "../../shared/src/loggers/ILogger.js";
export type { AttachmentFn } from "../../shared/src/attachments/Attachments.js";

export { TreeReport } from "./reports/TreeReport.js";
export { AnalyzeReport } from "./reports/AnalyzeReport.js";
export { DownloadReport } from "./reports/DownloadCountReport.js";
export { LicenseReport } from "./reports/LicenseReport.js";
export { LoopsReport } from "./reports/LoopsReport.js";
export { UpdateInfoReport } from "./reports/UpdateInfoReport.js";

export type { ILintTypes, ILintCheck } from "./reports/lint/LintRule.js";
export { createRule } from "./reports/lint/LintRule.js";
export { ValidateKey } from "./reports/lint/checks/ValidateKey.js";
