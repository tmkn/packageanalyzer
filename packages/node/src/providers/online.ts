import { QueryClient } from "@tanstack/query-core";

import {
    type IPackageMetadata,
    type IUnpublishedPackageMetadata
} from "../../../shared/src/npm.js";
import type { Url } from "../../../shared/src/reports/Validation.js";
import { AbstractPackageProvider } from "../../../shared/src/providers/provider.js";
import { downloadJson } from "../utils/requests.js";

//loads npm data from the web
export class OnlinePackageProvider extends AbstractPackageProvider {
    private readonly _queryClient = new QueryClient({
        defaultOptions: {
            queries: {
                retry: 3,
                retryDelay: 0, // with the default, the tests ran into timeouts, no delay yolo
                staleTime: Infinity // not really useful right now as caching is done via _cache
            }
        }
    });

    constructor(private readonly _url: Url) {
        super();
    }

    async getPackageMetadata(
        name: string
    ): Promise<IPackageMetadata | IUnpublishedPackageMetadata | undefined> {
        const cachedInfo = this._cache.get(name);

        if (typeof cachedInfo !== "undefined") {
            return cachedInfo;
        } else {
            const packageMetadata = await this._queryClient.fetchQuery({
                queryKey: ["package", name],
                queryFn: async ({ signal }) => {
                    const response = await downloadJson<
                        IPackageMetadata | IUnpublishedPackageMetadata
                    >(`${this._url}/${encodeURIComponent(name)}`, { signal });

                    if (response === null) {
                        console.error(
                            `Failed to fetch package metadata for "${name}" from ${this._url}, RETRYING...`
                        );
                        throw new Error(
                            `Failed to fetch package metadata for "${name}" from ${this._url}`
                        );
                    }

                    return response;
                }
            });

            return packageMetadata;
        }
    }
}

export const npmOnline = new OnlinePackageProvider(`https://registry.npmjs.com`);
