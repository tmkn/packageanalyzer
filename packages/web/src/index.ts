export { Visitor, getPackageVersionfromString } from "../../shared/src/visitors/visitor.js";
export type { PackageVersion } from "../../shared/src/visitors/visitor.js";
export { Package } from "../../shared/src/package/package.js";
export type { IPackageJsonProvider } from "../../shared/src/providers/provider.js";
export type { ILogger } from "../../shared/src/loggers/ILogger.js";
export type { AttachmentFn } from "../../shared/src/attachments/Attachments.js";
export type { ReportServiceConfig } from "../../shared/src/reports/ReportService.js";
