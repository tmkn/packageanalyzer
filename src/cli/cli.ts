import { Cli, Builtins } from "clipanion";

import { AnalyzeCommand } from "./analyzeCommand";
import { UpdateInfoCommand } from "./updateInfoCommand";
import { DownloadCommand } from "./downloadCommand";
import { LoopsCommand } from "./loopsCommand";
import { TreeCommand } from "./treeCommand";
import { NpmDumpCommand } from "./npmDumpCommand";
import { getVersion } from "./common";
import { NpmDumpLookupCreatorCommand } from "./npmLookupCreatorCommand";
import { LicenseCheckCommand } from "./licenseCommand";
import { ReportCommand } from "./reportCommand";
import { DependencyDumperCommand } from "./dependencyDumpCommand";

export const cli = new Cli({
    binaryLabel: `packageanalyzer`,
    binaryName: `pkga`,
    binaryVersion: getVersion()
});

//standard commands
cli.register(ReportCommand);
cli.register(AnalyzeCommand);
cli.register(UpdateInfoCommand);
cli.register(DownloadCommand);
cli.register(LoopsCommand);
cli.register(TreeCommand);
cli.register(LicenseCheckCommand);

//development niche commands
cli.register(NpmDumpCommand);
cli.register(NpmDumpLookupCreatorCommand);
cli.register(DependencyDumperCommand);

//built in commands
cli.register(Builtins.HelpCommand);
cli.register(Builtins.VersionCommand);