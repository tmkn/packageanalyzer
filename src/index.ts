export { Visitor, getPackageVersionfromString } from "./visitors/visitor";
export type { PackageVersion } from "./visitors/visitor";
export { Package } from "./package/package";
export { npmOnline, OnlinePackageProvider } from "./providers/online";
export type { IPackageJsonProvider } from "./providers/provider";
export { OraLogger } from "./loggers/OraLogger";
export type { ILogger } from "./loggers/ILogger";
export type { IDecorator } from "./extensions/decorators/Decorator";

export { TreeReport } from "./reports/TreeReport";
export { AnalyzeReport } from "./reports/AnalyzeReport";
export { DownloadReport } from "./reports/DownloadCountReport";
export { LicenseReport } from "./reports/LicenseReport";
export { LoopsReport } from "./reports/LoopsReport";
export { UpdateInfoReport } from "./reports/UpdateInfoReport";

export type { ILintTypes, ILintCheck } from "./reports/lint/LintRule";
export { createRule } from "./reports/lint/LintRule";
