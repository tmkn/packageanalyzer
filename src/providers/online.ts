import * as semver from "semver";

import { IPackageVersionProvider } from "./folder";
import {
    INpmPackage,
    INpmPackageVersion,
    PackageVersion,
    IUnpublishedNpmPackage,
    isUnpublished
} from "../npm";
import { downloadHttpJson } from "../requests";

export abstract class PackageProvider {
    abstract getPackageInfo(
        name: string
    ): Promise<INpmPackage | IUnpublishedNpmPackage | undefined>;
}

//loads npm data from the web
export class OnlinePackageProvider extends PackageProvider implements IPackageVersionProvider {
    private readonly _cache: Map<string, INpmPackage> = new Map();

    constructor(private _url: string, private _max = 3) {
        super();
    }

    /*setMaxConcurrent(newMax: number): void {
        this._max = newMax;
    }*/

    get size(): number {
        let size = 0;

        for (const [, info] of this._cache) {
            size += Object.keys(info.versions).length;
        }

        return size;
    }

    async getPackageInfo(name: string): Promise<INpmPackage | IUnpublishedNpmPackage | undefined> {
        const cachedInfo = this._cache.get(name);

        if (typeof cachedInfo !== "undefined") {
            return cachedInfo;
        } else {
            const data = await downloadHttpJson<INpmPackage>(
                `${this._url}/${encodeURIComponent(name)}`
            );

            return data === null ? undefined : data;
        }
    }

    async *getPackagesByVersion(
        modules: PackageVersion[]
    ): AsyncIterableIterator<INpmPackageVersion> {
        for (const [name, version] of modules) {
            yield this.getPackageByVersion(name, version);
        }
    }

    async getPackageByVersion(
        name: string,
        version: string | undefined = undefined
    ): Promise<INpmPackageVersion> {
        let info: INpmPackage | IUnpublishedNpmPackage | undefined = this._cache.get(name);

        if (!info) {
            info = await this.getPackageInfo(name);

            if (!info) {
                const _version: string = typeof version !== "undefined" ? `@${version}` : ``;
                throw new Error(`Couldn't get package "${name}${_version}"`);
            }

            if (isUnpublished(info)) {
                throw new Error(`Package "${name}" was unpublished`);
            }

            this._cache.set(name, info);
        }

        const allVersions: string[] = Object.keys(info.versions);
        const versionToResolve =
            typeof version !== "undefined" ? version : info["dist-tags"].latest;
        const resolvedVersion: string | null = semver.maxSatisfying(allVersions, versionToResolve);

        if (resolvedVersion === null) {
            throw new Error(`Couldn't resolve version ${version} for "${name}"`);
        }

        return info.versions[resolvedVersion];
    }
}

export const npmOnline = new OnlinePackageProvider(`http://registry.npmjs.com`);
