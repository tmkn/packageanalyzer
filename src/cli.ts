#!/usr/bin/env node

const pkg = require("./../../package.json");

import * as dayjs from "dayjs";

import { resolveFromFolder } from "./resolvers/folderResolver";
import { resolveFromName } from "./resolvers/nameResolver";
import { npmOnline } from "./providers/onlineProvider";
import {
    PackageAnalytics,
    LicenseSummary,
    VersionSummary,
    getNameAndVersion,
    groupPackagesByLicense
} from "./analyzer";

let commandFound = false;

process.argv.forEach((arg, i) => {
    if (arg === "-v" && !commandFound) {
        console.log(`Version ${pkg.version}`);
        commandFound = true;
    } else if (arg === "-h" && !commandFound) {
        showHelp();
        commandFound = true;
    } else if (arg === "-f" && !commandFound) {
        let folder = process.argv[i + 1];

        cliResolveFolder(folder);
        commandFound = true;
    } else if (arg === "-o" && !commandFound) {
        let name = process.argv[i + 1];

        cliResolveName(name);
        commandFound = true;
    }
});

if (!commandFound) {
    console.log(`No command found`);
    showHelp();
}

function showHelp(): void {
    console.log(`Package Analyzer ${pkg.version}`);
    console.log(`Options:`);
    console.log(
        `${`-f`.padEnd(10)} ${`Analyze a local folder`.padEnd(60)} e.g. "npa -f path/to/project"`
    );
    console.log(
        `${`-o`.padEnd(10)} ${`Analyze a package, version optional, default latest`.padEnd(
            60
        )} e.g. "npa -o typescript(@3.5.1)"`
    );
    console.log(`${`-v`.padEnd(10)} ${`Prints version`.padEnd(60)}`);
    console.log(`${`-h`.padEnd(10)} ${`Display help`.padEnd(60)}`);
}

async function cliResolveFolder(folder: string | undefined): Promise<void> {
    if (typeof folder === "undefined") {
        console.log(`Missing folder path\n`);
        showHelp();

        return;
    }

    try {
        let pa: PackageAnalytics = await resolveFromFolder(folder);

        printStatistics(pa);
    } catch (e) {
        //console.log(e);
    }
}

async function cliResolveName(pkgName: string | undefined): Promise<void> {
    if (typeof pkgName === "undefined") {
        console.log(`Missing package name\n`);
        showHelp();

        return;
    }

    try {
        const [name, version] = getNameAndVersion(pkgName);
        let pa: PackageAnalytics;

        if (typeof version === "undefined") pa = await resolveFromName(name, npmOnline);
        else pa = await resolveFromName([name, version], npmOnline);

        printStatistics(pa);
    } catch (e) {
        console.log(e);
    }
}

function printStatistics(pa: PackageAnalytics): void {
    const padding = 40;
    const paddingLeft = 4;

    console.log(`Statistics for ${pa.fullName}\n`);

    printPublished(pa, padding);
    printOldest(pa.oldest, padding);
    printNewest(pa.newest, padding);
    console.log(`${`Direct dependencies:`.padEnd(padding)}${pa.directDependencyCount}`);
    console.log(`${`Transitive dependencies:`.padEnd(padding)}${pa.transitiveDependenciesCount}`);
    printDistinctDependencies(pa.distinctByNameCount, pa.distinctByVersionCount, paddingLeft);
    printMostReferred(pa.mostReferred, padding);
    printMostDependencies(pa.mostDependencies, padding);
    printMostVersion(pa.mostVersions, padding, paddingLeft);
    printLoops(pa.loops, padding, paddingLeft);
    printLicenseInfo(pa.licenses, paddingLeft);
}

function printNewest(newest: PackageAnalytics | undefined, padding: number): void {
    if (newest && newest.published) {
        console.log(
            `${`Newest:`.padEnd(padding)}${
                newest.fullName
            } - ${newest.published.toUTCString()} ${daysAgo(newest.published)}`
        );
    }
}

function printOldest(oldest: PackageAnalytics | undefined, padding: number): void {
    if (oldest && oldest.published) {
        console.log(
            `${`Oldest:`.padEnd(padding)}${
                oldest.fullName
            } - ${oldest.published.toUTCString()} ${daysAgo(oldest.published)}`
        );
    }
}

function daysAgo(date: number | Date): string {
    return `(${dayjs(new Date()).diff(date, "day")} days ago)`;
}

function printPublished(pa: PackageAnalytics, padding: number): void {
    if (!pa.published) return;

    console.log(
        `${`Published:`.padEnd(padding)}${pa.published.toUTCString()} ${daysAgo(pa.published)}`
    );
}

function printDistinctDependencies(
    byName: number,
    byNameAndVersion: number,
    paddingLeft: number
): void {
    console.log(`Distinct dependencies:`);
    console.log(`${``.padStart(paddingLeft)}${byName}: distinct name`);
    console.log(`${``.padStart(paddingLeft)}${byNameAndVersion}: distinct name and version`);
}

function printLoops(loops: PackageAnalytics[], padding: number, paddingLeft: number): void {
    console.log(`${`Loops:`.padEnd(padding)}${loops.length}`);

    if (loops.length > 0) {
        let [first] = loops.map(l => l.pathString).sort();

        console.log(`    e.g. ${first}`);

        if (loops.length > 1)
            console.log(`${``.padStart(paddingLeft)}+${loops.length - 1} additional loops`);
    }
}

function printMostDependencies(pa: PackageAnalytics, padding: number): void {
    console.log(
        `${`Most direct dependencies:`.padEnd(padding)}"${pa.name}": ${pa.directDependencyCount}`
    );
}

function printMostReferred(arg: [string, number], padding: number): void {
    console.log(`${`Most referred package:`.padEnd(padding)}"${arg[0]}": ${arg[1]}x`);
}

function printMostVersion(mostVerions: VersionSummary, padding: number, paddingLeft: number): void {
    console.log(`${`Package(s) with most versions:`.padEnd(padding)}`);

    for (const [name, versions] of mostVerions) {
        console.log(`${``.padStart(paddingLeft)}${name} [${[...versions].join(", ")}]`);
    }
}

function printLicenseInfo(allLicenses: LicenseSummary, paddingLeft: number): void {
    let summary = new Map<string, number>();
    let longestLine = 0;

    for (const [, licenses] of allLicenses) {
        for (const [, license] of licenses) {
            let specificLicense = summary.get(license);

            if (!specificLicense) {
                summary.set(license, 1);
            } else {
                summary.set(license, specificLicense + 1);
            }

            if (license.length > longestLine) longestLine = license.length;
        }
    }

    console.log(`Licenses:`);
    for (const { license, names } of groupPackagesByLicense(allLicenses)) {
        let threshold = 4;
        let samples: string[] = names.slice(0, threshold);

        if (names.length >= threshold)
            console.log(
                `${``.padStart(paddingLeft)}${license} - [${samples.join(", ")}, +${names.length -
                    threshold} more]`
            );
        else console.log(`${``.padStart(paddingLeft)}${license} - [${samples.join(", ")}]`);
    }

    /*let longestLine = 0;

    for(const [name, licenses] of allLicenses) {
        for(const [version, license] of licenses) {
            let fullName = `${name}@${version}`;

            if(fullName.length > longestLine)
                longestLine = fullName.length;
        }
    }

    for(const [name, licenses] of allLicenses) {
        for(const [version, license] of licenses) {
            console.log(`${`${name}@${version}`.padEnd(longestLine + 1)}${license}`.padStart(8));
        }
    }*/
}
