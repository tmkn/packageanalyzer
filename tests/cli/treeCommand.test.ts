import * as path from "path";
import { PassThrough } from "stream";

import { BaseContext } from "clipanion";

import { cli } from "../../src/cli";
import { OnlinePackageProvider } from "../../src/providers/online";
import { createMockNpmServer, MockNpmServer } from "../server";

describe(`Tree Command`, () => {
    const mockContext: BaseContext = {
        stdin: process.stdin,
        stdout: new PassThrough(),
        stderr: new PassThrough()
    };

    let server: MockNpmServer;
    let provider: OnlinePackageProvider;

    beforeAll(async () => {
        server = await createMockNpmServer();
        provider = new OnlinePackageProvider(`http://localhost:${server.port}`);
    });

    test(`--package --type`, async () => {
        const command = cli.process([
            `tree`,
            `--package`,
            `react@16.8.1`,
            `--type`,
            `dependencies`
        ]);

        expect.assertions(0);
        command.context = mockContext;
        await command.execute();
    });

    test(`--folder --type`, async () => {
        const command = cli.process([
            `tree`,
            `--folder`,
            path.join("tests", "data", "testproject1"),
            `--type`,
            `dependencies`
        ]);

        expect.assertions(0);
        command.context = mockContext;

        await command.execute();
    });

    afterAll(() => server.close());
});
