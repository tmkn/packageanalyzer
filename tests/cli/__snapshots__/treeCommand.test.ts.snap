// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tree Command --folder --type 1`] = `
Array [
  "testproject1@1.0.0 (13 dependencies)",
  "└── react@16.8.6 (12 dependencies)",
  "    ├── loose-envify@1.4.0 (1 dependencies)",
  "    │   └── js-tokens@4.0.0 (0 dependencies)",
  "    ├── object-assign@4.1.1 (0 dependencies)",
  "    ├── prop-types@15.7.2 (4 dependencies)",
  "    │   ├── loose-envify@1.4.0 (1 dependencies)",
  "    │   │   └── js-tokens@4.0.0 (0 dependencies)",
  "    │   ├── object-assign@4.1.1 (0 dependencies)",
  "    │   └── react-is@16.8.6 (0 dependencies)",
  "    └── scheduler@0.13.6 (3 dependencies)",
  "        ├── loose-envify@1.4.0 (1 dependencies)",
  "        │   └── js-tokens@4.0.0 (0 dependencies)",
  "        └── object-assign@4.1.1 (0 dependencies)",
  "",
]
`;

exports[`Tree Command --package --type 1`] = `
Array [
  "react@16.8.1 (12 dependencies)",
  "├── loose-envify@1.4.0 (1 dependencies)",
  "│   └── js-tokens@4.0.0 (0 dependencies)",
  "├── object-assign@4.1.1 (0 dependencies)",
  "├── prop-types@15.7.2 (4 dependencies)",
  "│   ├── loose-envify@1.4.0 (1 dependencies)",
  "│   │   └── js-tokens@4.0.0 (0 dependencies)",
  "│   ├── object-assign@4.1.1 (0 dependencies)",
  "│   └── react-is@16.13.1 (0 dependencies)",
  "└── scheduler@0.13.6 (3 dependencies)",
  "    ├── loose-envify@1.4.0 (1 dependencies)",
  "    │   └── js-tokens@4.0.0 (0 dependencies)",
  "    └── object-assign@4.1.1 (0 dependencies)",
  "",
]
`;
